drafts/refactoring-legacy-code-with-tiny-types-in-java.md

```diff
@@ -0,0 +1,237 @@
+  }
+}
+```
+
+Before diving into refactoring the legacy code, it’s best to have some kind of safety net in place.
+Instead of writing tons of tiny unit tests, we can go for a broader 'approval' test that gives us quick feedback.
+Approval tests verify if the output from our code matches an 'approved' result.
+Unlike the usual tests that assert specific values or behaviors, these tests capture the output and compare it to a pre-approved baseline or 'golden master.'
+If the current output strays from what’s approved, the test fails, letting us decide whether to approve the new output or tweak the code.
+
+In the Gilded Rose Kata, we can implement an approval test by printing the `sellIn` and `quality` of each item to a file at the end of each day.
+This approach allows us to capture the state of our inventory over time and compare it against an expected output.
+
+Once setup, our approval test will verify that the output remains consistent as we modify the code.
+At this point, we are ready to start refactoring.
+
+The [Gilded Rose requirements](https://github.com/emilybache/GildedRose-Refactoring-Kata/blob/main/GildedRoseRequirements.md) state that:
+
+* The `Quality` of an item is never negative
+* The `Quality` of an item is never more than `50`
+
+The original `Item` class in the exercise serves as a mutable data structure, and the codebase includes a great deal of bound-checks to ensure that the requirements aren't violated.
+
+```java
+if (items[i].quality < 50) {
+  items[i].quality = items[i].quality + 1;
+}
+
+if (items[i].quality > 0) {
+  items[i].quality = items[i].quality - 1;
+}
+```
+
+A much better design involves introducing a tiny type to encompass the constraints.
+In this instance, we use a Java Record.
+[Java Records](https://openjdk.org/jeps/395), introduced in Java 16, provide a concise way to define immutable data classes.
+A record automatically generates the boilerplate code typical of simple data holding classes - such as constructors, getters, `equals()`, `hashCode()`, and `toString()` - while indicating that the class serves solely to store data.
+
+```java
+public record BoundedQuality(int value) {
+  public BoundedQuality {
+    if (value < 0 || value > 50) {
+      throw new IllegalArgumentException("quality must be between 0-50");
+    }
+  }
+
+  public static UnaryOperator<BoundedQuality> adjustQuality(int adjustment) {
+    return it -> new BoundedQuality(Math.max(0, Math.min(50, it.value + adjustment)));
+  }
+}
+```
+
+Whenever a method receives a `BoundedQuality`, we can confidently assert that the data's integrity remains intact, allowing us to remove unnecessary defensive checks.
+Additionally, the `adjust` method enables safe manipulation of the quality amount.
+Moreover, if we attempt to create a `BoundedQuality` with invalid data, such as a negative amount, the system fails fast.
+
+Before we introduced the `BoundedQuality` type, we modeled both the `quality` and `sellIn` amounts as primitive integers.
```

Comment:
I was expecting to read something about sellIn .  I appreciate this is the same solution as quality, ie: its own java record.  Maybe mention this?  Or just focus on quality and remove sellIn from the article?